### DISCUSSION TITLE
door intercom

### DISCUSSION THREAD

**Discussion by gemu2015** [arendst/Tasmota#16226] [Show and tell]
this describes an audio intercom with push to talk interface.

it may be used for a door intercom,  or baby phone or similar

Hardware requirements:
any ESP32 should work (except C3), no PSRAM needed.
I2S Audio DAC with amplifier https://www.ebay.de/itm/165446200440?hash=item26855cac78:g:A8EAAOSwDspiY5o4
I2S microphone https://www.ebay.de/itm/173974875385?hash=item2881b5ecf9:g:B-cAAOSwZbRdVhJf

connect Audio DAC and microphone to any suitable ESP32 pins, microphone VCC must not be higher than 3.3 V. connect L/R pin to GND, connect an (8 Ohm) speaker to the DAC
and set the appropriate I2S settings in Tasmota GPIO WebUI.
I2S_BLK,I2S_WS,I2S_DOUT,I2S_DIN

you need 2 of these ESP devices, one for the door and one in house.

these are the console cmds to setup the bridge

// set slave ip, the ip of the door device
i2sbridge 192.168.188.105
// swap microphone channels, 6 swap on, 7 = swap off
i2sbridge 6
// set master mode, 4 = master, 5 = slave
i2sbridge 4
// set push to talk pin (here pin 13)
// when set pushing sets write mode, releasing sets read mode
i2sbridge p13
; start bridge  1 = read from slave, 2 = write to slave, 3 = loopback, 0 = off
i2sbridge 1
; set micro gain factor G = 1-50
i2smgain G

slave does not need any preset, except mgain or mic swap when desired
 
currently settings are not persistent and have to be preset at boot time by rule or script

here is a script example to setup a master device

\>D
once=0
res=0
p:micgain=0
micrec=0
bridge=0

\>BS

\>F
if chg[micgain]>0 {
	->i2smgain %0micgain%
}

\>S
if wifis>0
and once==0 {
	; set slave ip
	=>i2sbridge 192.168.188.105
	;swap microphone source
	=>i2sbridge 6
	; set to master mode
	=>i2sbridge 4
	; set ptt pin
	=>i2sbridge p13
	; set mic gain
	=>i2smgain %0micgain%
	once=1
}

if chg[bridge]>0 {
	if bridge>0 {
		=>i2sbridge 1
	} else {
		=>i2sbridge 0
	}
}

\>WS
sl(1 50 micgain "1" "Micro gain : %0micgain%" "50")
bu(bridge  "Stop audio bridge" "Start audio bridge")

defines for user_config_override:
\#define USE_SHINE
\#define USE_I2S_AUDIO
\#define USE_I2S_MIC
\#define I2S_BRIDGE
\#define MICSRATE 16000
\#define MIC_CHANNELS 2

**Comment by dcast78** [comment]
Trying to use your user_config_override

#define USE_SHINE
#define USE_I2S_AUDIO
#define USE_I2S_MIC
#define I2S_BRIDGE
#define MICSRATE 16000
#define MIC_CHANNELS 2

I have 2 problems

1. USE_I2S_MIC produce this error
`tasmota/tasmota_xdrv_driver/xdrv_42_1_i2s_mp3mic.ino:128:10: fatal error: layer3.h: No such file or directory`

2. USE_I2S_AUDIO no error but command Unknown on console
`19:16:34.283 CMD: i2sbridge 6`
`19:16:34.289 RSL: RESULT = {"Command":"Unknown"}` 

Environment:
- Visual studio code with platformio. 
- Fresh clone of tasmota repo.
- Correct upload and working tasmota firmware

Any idea?

Thanks

**Comment by gemu2015** [comment]
you must add in linker section lib_extra_dirs 
lib/libesp32_audio

Then define the i2s pins in tasmota GUI and watch the console on reboot

**Reply by dcast78** [reply]
Ok
also added lib/libesp32

now my lib_extra_dirs is

```
lib_extra_dirs          = ${common.lib_extra_dirs}
                        lib/libesp32
                        lib/libesp32_audio
```

all commands are accepted.

I2s work correctly with i2srec and i2splay.

To add script this is needed

```
#ifndef USE_SCRIPT
#define USE_SCRIPT  // adds about 17k flash size, variable ram size
#endif
#ifdef USE_RULES
#undef USE_RULES
#endif 
```
At the moment intercom is not working, but I have made some steps forward.

A question... You have used only one I2S bus or two I2S bus, one for ti he mic and one for the speaker?

**Reply by gemu2015** [reply]
no, you must use one i2s bus for mic and speaker, either i2s bus number 1 or 2

watch the web console in master and slave, this should give hints what is going on

**Comment by blakadder** [comment]
I'm trying to use this feature but `i2sbridge` and also `i2srec` ans `i2sstream` is crashing Tasmota on multiple devices (M5StickC, ESP-S2 with microphone, XIAO ESPS3 Sense.
This is the error:
```
Guru Meditation Error: Core  1 panic'ed (IntegerDivideByZero). Exception was unhandled.

Core  1 register dump:
PC      : 0x4019a7bb  PS      : 0x00060d30  A0      : 0x8019aed4  A1      : 0x3ffd2df0  
A2      : 0x00000000  A3      : 0x00000000  A4      : 0x3ffeea3c  A5      : 0x00000000  
A6      : 0x3ffeea3c  A7      : 0x3ffc7ea8  A8      : 0x00000100  A9      : 0x00000002  
A10     : 0x00000000  A11     : 0x00000000  A12     : 0x00000000  A13     : 0xffffefff  
A14     : 0xffffffbf  A15     : 0x00000008  SAR     : 0x00000020  EXCCAUSE: 0x00000006  
EXCVADDR: 0x00000000  LBEG    : 0x4008c5ac  LEND    : 0x4008c5b7  LCOUNT  : 0xffffffff  


Backtrace: 0x4019a7b8:0x3ffd2df0 0x4019aed1:0x3ffd2e30 0x4016422f:0x3ffd2e60 0x401642a2:0x3ffd2ed0 0x40165a4f:0x3ffd2f00 0x4014b211:0x3ffd2f50 0x4010b5da:0x3ffd2f70 0x4010ca51:0x3ffd2f90 0x4010e215:0x3ffd2fc0 0x40164402:0x3ffd3030 0x4014e8c5:0x3ffd3050 0x40187756:0x3ffd3070 0x40196d81:0x3ffd3090 0x400f6d3f:0x3ffd3100
```

**Reply by gemu2015** [reply]
did you apply these defines ?

#define MP3_MIC_STREAM
#define USE_I2S_MIC
#define USE_SHINE
#define USE_I2S_AUDIO
#define I2S_BRIDGE

and you need PSRAM for  STREAM, and REC cause MP3 Encoder needs much RAM

and you must setup in TASMOTA GUI
I2S_BLK,I2S_WS,I2S_DOUT,I2S_DIN

i did successfully test with ESP32 and ESP32-S3 not S2

**Comment by fdaubercy** [comment]
Hello everyone.
I am working on using the ESP32-P4 devkit: 
See document below
[GPIO Pins - ESP32-P4-Module-DEV-KIT.pdf](https://github.com/user-attachments/files/24287936/GPIO.Pins.-.ESP32-P4-Module-DEV-KIT.pdf).

<img width="486" height="650" alt="image" src="https://github.com/user-attachments/assets/3633ac7f-2dff-41f8-b3e9-5ac6a15d604b" />
<img width="913" height="755" alt="image" src="https://github.com/user-attachments/assets/0986aaa6-c248-45fd-b714-90c658915cc1" />

The pins are configured as follows:

- I2S DIN: GPIO9
- I2S WS: GPIO10
- I2S DOut: GPIO11
- I2S BClk: GPIO12
- I2S MClk: GPIO13
- Hi Output: GPIO53 (PA activation).

I increased the Gain to 100 (i2sconfig Gain 100).

This doesn't work, neither with an MP3 file nor with the 'I2STime' command.

I compiled with the following settings in the 'user_config_override.h' file:

- USE_I2S_ALL
- USE_I2S
- USE_I2S_AUDIO
- USE_I2S_MIC
- USE_SHINE
- MP3_MIC_STREAM
- USE_I2S_AUDIO_BERRY
- USE_I2S_AAC
- USE_I2S_OPUS
- USE_I2S_SAY_TIME
- USE_I2S_RTTTL

Can you help me, please?

**Reply by gemu2015** [reply]
this board uses a codec ES8311 as audio interface. there is currently no support for this chip.
@Staars may help

**Reply by sfromis** [reply]
Staars does have a Gist with a Berry driver for ES8311 "may work with the ESP32P4-EV too"
https://gist.github.com/Staars/48d784543dba982dcdf51b0de0636af3
Not having tried it, I can't say much about completeless or how integrated it is.

Surprisingly, there exists a I2S driver mentioning ES8311, albeit "for ESP32-S3 box and box lite"
https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_42_4_i2s_codecs.ino
Again, outside of what I've played with.

**Reply by Staars** [reply]
The linked GIST should work as a base, but the power enable pin is different there with GPIO 20. Just change the hardcoded setting for GPIO53 like so:

```
    # PA Power Control
    def pa_power(enable)
        log(f"ES8311: Setting PA GPIO 53 to {enable}",4)
        import gpio
        if enable
            gpio.digital_write(53, 1)
        else
            gpio.digital_write(53, 0)
        end
    end
```

**Reply by fdaubercy** [reply]
I ran the following command in the terminal: 'I2SGain 100'.
Then I loaded the driver using the Berry command:'load("/guition_es8311.be")' after replacing gpio20 with gpio53 in the Berry code.
Finally, I still hear nothing when I try the 'I2STime' command.

**Reply by Staars** [reply]
On my board DIN and DOUT have the opposite order, meaning:
DOUT = 9
DIN = 11

Maybe you can try that.
At least I can confirm that i2ssay and i2stime do work with a fresh build from today on my P4.

**Reply by fdaubercy** [reply]
Thank you so much, it works with this new GPIO9 and 11 configuration.
**How did you configure GPIO53?**

For me, the sound is still very low despite the 'I2CGain 100' command.
For this new test, I didn't run the BERRY script 'guition_es8311.be'. 
**Do you use it?**

To make the sound more audible, **do I need to modify the I2S parameters further with the 'I2SConfig' command?**

**Reply by Staars** [reply]
As I wrote above: My power enable pin is 20 and not 53.

When you do not run the Berry script, the codec is in an undefined state, which may be still workable after a warm start.
It is definitely my advice to use the script.

You can crank up the codec volume to 100% with:
codec  = ES8311()
codec.set_volume(100)

(Or just edit the bottom lines of my script)

**Reply by fdaubercy** [reply]
Yes, it works with this script.
The sound is audible but poor quality with crackling.
For my module, GPIO53 is indeed the one that activates the audio amplifier.

**Reply by sfromis** [reply]
If you want to check what sound quality the chip can deliver, better play a MP3 file instead of using  the voice generation commands `I2SSay` and `I2STime`, which are not great sound quality already.
